From 7fb5b6727d6399f93363e914b5dc1d98f59fd4eb Mon Sep 17 00:00:00 2001
From: Rander Wang <rander.wang@linux.intel.com>
Date: Thu, 4 Jul 2019 15:03:48 +0800
Subject: [PATCH] FROMGIT: ASoC: SOF: hda: fix stream id setting

snd_hdac_ext_link_clear_stream_id maps stream id to link output,
which is for playback, not capture.

Tested on Whiskey Lake platform.

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>
---
 sound/soc/sof/intel/hda-dai.c | 6 ++++--
 sound/soc/sof/intel/hda-dsp.c | 5 +++++
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index 5b83387af6ca..4c0d7880e31c 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -330,7 +330,8 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 		if (ret < 0)
 			return ret;
 		stream_tag = hdac_stream(link_dev)->stream_tag;
-		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+			snd_hdac_ext_link_clear_stream_id(link, stream_tag);
 
 		/* fallthrough */
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
@@ -372,7 +373,8 @@ static int hda_link_hw_free(struct snd_pcm_substream *substream,
 		return -EINVAL;
 
 	stream_tag = hdac_stream(link_dev)->stream_tag;
-	snd_hdac_ext_link_clear_stream_id(link, stream_tag);
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
 	snd_hdac_ext_stream_release(link_dev, HDAC_EXT_STREAM_TYPE_LINK);
 	snd_soc_dai_set_dma_data(dai, substream, NULL);
 	link_dev->link_prepared = 0;
diff --git a/sound/soc/sof/intel/hda-dsp.c b/sound/soc/sof/intel/hda-dsp.c
index edb6d56c77a4..6665e006770c 100644
--- a/sound/soc/sof/intel/hda-dsp.c
+++ b/sound/soc/sof/intel/hda-dsp.c
@@ -501,6 +501,11 @@ int hda_dsp_set_hw_params_upon_resume(struct snd_sof_dev *sdev)
 			if (!link)
 				return -EINVAL;
 			stream_tag = hdac_stream(stream)->stream_tag;
+
+			if (hdac_stream(stream)->direction ==
+				SNDRV_PCM_STREAM_CAPTURE)
+				continue;
+
 			snd_hdac_ext_link_clear_stream_id(link, stream_tag);
 		}
 #endif
-- 
2.17.1

